web/website/content/posts/2023-03-14-pi-day.md

```diff
@@ -0,0 +1,370 @@
+of the first examples when recursion is introduced:
+
+```elm=
+from_text format:json '[{"a":1, "b":1}]'
+loop (
+    select [a=b, b=s"a"+b]
+)
+select a
+take 7
+```
+
+which produces the first 7 Fibonnaci numbers.
+
+| a   |
+| --- |
+| 1   |
+| 1   |
+| 2   |
+| 3   |
+
+|5 |8| |13|
+
+You might have noticed that we didn't actually include a `filter` in our loop
+this time. Instead we relied on the fact that DuckDB produces results lazily and
+since we only took 7 numbers it only produced what we needed.
+
+Now, let's set ourselves a bigger challenge.
+
+## Calculating the digits of pi
+
+March 14th is written on American calendars as 3/14 which reminds us of the
+first three digits of Pi=3.1415926535... . Therefore this day is commonly known
+as Pi day. In order to celebrate Pi Day 2023 and the recent release of `loop` in
+PRQL 0.6, why don't we try to calculate the digits of Pi in PRQL!
+
+For our query engine we will use [DuckDB](https://duckdb.org/) because it is
+really fast and has many modern features that make it ideally suited for the
+kinds of analytical queries that PRQL is targeting. There is even a
+[duckdb-prql](https://github.com/ywelsch/duckdb-prql) extension which allows you
+to write PRQL queries right inside DuckDB!
+
+We follow the algorithms in the paper
+["Unbounded Spigot Algorithms for the Digits of Pi" by Jeremy Gibbons](https://www.cs.ox.ac.uk/people/jeremy.gibbons/publications/spigot.pdf),
+in particular Rabinowitz and Wagon's Spigot Algorithm. While it is not the most
+efficient algorithm presented in the paper, it has the advantage that it uses
+only standard data types while the more efficient algorithms rely on unbounded
+integer types which are not available in many database systems.
+
+For my implementation, I adapted the Python implementation by John Burkardt
+found
+[here](https://people.sc.fsu.edu/~jburkardt/py_src/pi_spigot/pi_spigot.py). A
+bit of care had to be taken in porting the implementation to PRQL, as PRQL like
+SQL, is stateless so I introduced another state variable `k` to track the
+position inside the body of the loop.
+
+The resulting PRQL query is the following:
+
+```prql
+prql target:sql.duckdb
+
+let config = (
+    from_text format:json '[{"n":50}]'
+    derive [
+        len = (10*n)/3,
+        calc_steps = 1+4,
+        loop_steps = len + calc_steps,
+        ]
+)
+
+from config
+select [
+    s = -1,
+    n,
+    len,
+    loop_steps,
+    j = 0,
+    k = 0,
+    q = 0,
+    a = s"[2 for i in range(len)]",
+    nines = 0,
+    predigit = 0,
+    output = '',
+]
+loop (
+    filter s"s" < s"n"*s"loop_steps" + 1
+    select [
+        s = s"s" + 1,
+        s"n",
+        s"len",
+        loop_steps,
+        j = case [s"k"==0 => s"j"+1, true => s"j"],
+        k = (s"k"+ 1) % s"loop_steps",
+        q = case [
+            s"k"==0 => 0,
+            1 <= s"k" and s"k" <= s"len" => (10*s"a[k]" + s"q"*(len-s"k"+1)) / (2*(len-s"k") + 1),
```

Comment:
Gosh, there are a lot of s-strings... Definitely worth explaining this I think, and thinking about how we can improve it

Reply:
It comes down to those select semantics. It's very convenient for normal queries (I guess 90% of queries) but it doesn't work for iteration. I learned a lot from this implementations.

Based on that recent type system RFC, I think the same probably holds for the rest of the language - it's been a super useful platform for experimentation and has been useful for doing real work from the get go. As we think about defining a type system we should try to incorporate all these learnings and try to figure out how we can support as many of them without introducing too much overhead.

Reply:
We cannot have a post that needs to contain hacks around semantics we defined. Can I try to refactor it somehow?

Reply:
You're welcome to try but I don't think it's possible. Try for the simple Fibonnaci example first and see if you can do it there because that illustrates the semantic issue.

Reply:
For the pi calculation, we could possibly do away with the s-strings by introducing more levels of the k counter. However then the compiler generates more and more levels of CTEs and nesting which makes manually fixing the query intractable.

Reply:
I will try an idea for pushing the k checks into a function which would hide at least some of the s-strings.

Reply:
I think that function refactor is working. Makes the logic also much clearer and highlights PRQL better. Give me 30 min.

Reply:
Done. Check out the new query! Also nicely highlights the power of functions in PRQL!